api.js

```diff
@@ -720,6 +725,458 @@ authenticate.clientCache  = clientCache;
+        // exploit... Besides we have plenty of obscurity to protect us here :)
+        client = limitClientWithExt(client, ext);
+      }
+      callback(null, {
+        clientToken:  client.clientId,
+        key:          client.accessToken,
+        algorithm:    'sha256',
+        scopes:       client.scopes
+      });
+    }).catch(callback);
+  };
+  return function(req) {
+    return new Promise(function(accept) {
+      var authenticated = function(err, credentials, artifacts) {
+        var result = null;
+        if (err) {
+          var message = "Unknown authorization error";
+          if (err.output && err.output.payload && err.output.payload.error) {
+            message = err.output.payload.error;
+            if (err.output.payload.message) {
+              message += ": " + err.output.payload.message;
+            }
+          } else if(err.message) {
+            message = err.message;
+          }
+          result = {
+            message:  '' + message
+          };
+        } else {
+          result = {
+            scheme:   'hawk',
+            scopes:   credentials.scopes
+          };
+          if (artifacts.hash) {
+            result.hash = artifacts.hash;
+          }
+        }
+        return accept(result);
+      };
+      if (req.authorization) {
+        hawk.server.authenticate({
+          method:           req.method,
+          url:              req.resource,
+          host:             req.host,
+          port:             req.port,
+          authorization:    req.authorization
+        }, function(clientId, callback) {
+          var ext = undefined;
+
+          // Parse authorization header for ext
+          var attrs = hawk.utils.parseAuthorizationHeader(
+            req.authorization
+          );
+          // Extra ext
+          if (!(attrs instanceof Error)) {
+            ext = attrs.ext;
+          }
+
+          // Get credentials with ext
+          loadCredentials(clientId, ext, callback);
+        }, {
+          // Not sure if JSON stringify is not deterministic by specification.
+          // I suspect not, so we'll postpone this till we're sure we want to do
+          // payload validation and how we want to do it.
+          //payload:      JSON.stringify(req.body),
+
+          // We found that clients often have time skew (particularly on OSX)
+          // since all our services require https we hardcode the allowed skew
+          // to a very high number (15 min) similar to AWS.
+          timestampSkewSec: 15 * 60,
+
+          // Provide nonce manager
+          nonceFunc:    options.nonceManager
+        }, authenticated);
+      } else {
+      // If there is no authorization header we'll attempt a login with bewit
+        hawk.uri.authenticate({
+          method:           req.method.toUpperCase(),
+          url:              req.resource,
+          host:             req.host,
+          port:             req.port,
+          authorization:    req.authorization
```

Comment:
Also, doesn't `hawk.uri.authenticate` need a `query` property on the request object it receives?